### PR TITLE
o/hookstate/ctlcmd: introduce --user, --system and --users switches for snap service operations

### PR DESCRIPTION
**Pull request by Meulengracht**
This is the snapctl version of https://github.com/snapcore/snapd/pull/13368 and based on top of it to reuse some of the code from that PR.

### CODE REVIEW

File: overlord/hookstate/ctlcmd/service_scope.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	User   bool   `long:"user"`
+	Users  string `long:"users"`
+}
+
+func (us *userAndScopeOptions) validateScopes() error {
+	switch {
+	case us.System && us.User:
+		return fmt.Errorf("--system and --user cannot be used in conjunction with each other")
+	case us.Users != "" && us.User:
+		return fmt.Errorf("--user and --users=all cannot be used in conjunction with each other")
```
Comment (Meulengracht) [inline overlord/hookstate/ctlcmd/service_scope.go:43]: change this to --users to match the previous PR

File: overlord/hookstate/ctlcmd/service_scope.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	"fmt"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// userAndScopeOptions represents shared options between service operations
+// that change the scope of services affected. This should more or less
+// match what's in cmd/snap.
+type userAndScopeOptions struct {
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/service_scope.go:32]: we have the same code in cmd mostly?  we have client/clientutil that is used already here to possibly share such things

File: overlord/hookstate/ctlcmd/services_test.go [hunk 3]
```diff
@@ -309,6 +309,70 @@ func (s *servicectlSuite) TestRestartCommand(c *C) {
...
+		}, "")
+		checkInvocation(c, names, []string{"--user"}, &servicestate.Instruction{
+			Action: c,
+			Names:  names,
+			Scope:  []string{"user"},
+			Users: client.UserSelector{
+				Selector: client.UserSelectionSelf,
+			},
+		}, "")
+		checkInvocation(c, names, []string{"--users=all"}, &servicestate.Instruction{
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/services_test.go:353]: should we check --system --users=all as well?

**Review by pedronis — COMMENTED** [review]
thank you, couple of comments

File: overlord/hookstate/ctlcmd/service_scope.go (hunk 2 above)
Comment (Meulengracht) [reply overlord/hookstate/ctlcmd/service_scope.go]: Moved the code into clientutil with dedicated tests

File: overlord/hookstate/ctlcmd/services_test.go (hunk 3 above)
Comment (Meulengracht) [reply overlord/hookstate/ctlcmd/services_test.go:365]: yes absolutely, thanks for reminding me!

**Review by bboozzoo — APPROVED** [review]
LGTM

**Pull request merged**